examples/pytorch/gat/train.py

```diff
@@ -0,0 +1,271 @@
+import time
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl import DGLGraph
+from dgl.data import register_data_args, load_data
+import dgl.function as fn
+
+class GraphAttention(nn.Module):
+    def __init__(self,
+                 g,
+                 in_dim,
+                 out_dim,
+                 num_heads,
+                 feat_drop,
+                 attn_drop,
+                 alpha,
+                 residual=False):
+        super(GraphAttention, self).__init__()
+        self.g = g
+        self.num_heads = num_heads
+        self.fc = nn.Linear(in_dim, num_heads * out_dim, bias=False)
+        if feat_drop:
+            self.feat_drop = nn.Dropout(feat_drop)
+        else:
+            self.feat_drop = None
+        if attn_drop:
+            self.attn_drop = nn.Dropout(attn_drop)
+        else:
+            self.attn_drop = None
+        self.attn_l = nn.Parameter(torch.Tensor(size=(num_heads, out_dim, 1)))
+        self.attn_r = nn.Parameter(torch.Tensor(size=(num_heads, out_dim, 1)))
+        nn.init.xavier_normal_(self.fc.weight.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_l.data, gain=1.414)
+        nn.init.xavier_normal_(self.attn_r.data, gain=1.414)
+        self.leaky_relu = nn.LeakyReLU(alpha)
+        self.residual = residual
+        if residual:
+            if in_dim != out_dim:
+                self.residual_fc = nn.Linear(in_dim, num_heads * out_dim, bias=False)
+                nn.init.xavier_normal_(self.fc.weight.data, gain=1.414)
+            else:
+                self.residual_fc = None
+
+    def forward(self, inputs):
+        # prepare
+        h = inputs
+        if self.feat_drop:
+            h = self.feat_drop(h)
+        ft = self.fc(h).reshape((h.shape[0], self.num_heads, -1))
+        head_ft = ft.transpose(0, 1)
+        a1 = torch.bmm(head_ft, self.attn_l).transpose(0, 1)
+        a2 = torch.bmm(head_ft, self.attn_r).transpose(0, 1)
+        self.g.ndata.update({'ft' : ft, 'a1' : a1, 'a2' : a2})
+        # 1. compute edge attention
+        self.g.apply_edges(self.edge_attention)
+        # 2. compute two results, one is the node features scaled by the dropped,
+        # unnormalized attention values. Another is the normalizer of the attention values.
+        self.g.update_all([fn.src_mul_edge('ft', 'a_drop', 'ft'), fn.copy_edge('a', 'a')],
+                          [fn.sum('ft', 'ft'), fn.sum('a', 'z')])
+        # 3. apply normalizer
+        ret = self.g.ndata['ft'] / self.g.ndata['z']
+        # 4. residual
+        if self.residual:
+            if self.residual_fc:
+                ret = self.residual_fc(h) + ret
+            else:
+                ret = h + ret
+        return ret
+
+    def edge_attention(self, edges):
+        # an edge UDF to compute unnormalized attention values from src and dst
+        a = self.leaky_relu(edges.src['a1'] + edges.dst['a2'])
+        a = torch.exp(a).clamp(-10, 10)  # use clamp to avoid overflow
```

Comment:
A caveat is that the softmax is technically still not the "correct and numerically-stable" softmax.  For example, if the pre-softmax scores are `[5, 10, 5, 5]` then the results would become, since here it just clips the whole thing to have a maximum absolute value of 10.

We should leave a comment there.

Reply:
I see. Is this a reasonable approximation? I found in the torch implementation
https://github.com/Diego999/pyGAT/blob/3301fd68cbf349925d7458d9d47766e48021cf78/layers.py#L111
They even use `torch.exp(-leaky_relu(...))`. Not sure whether this is equivalent.

Reply:
`-leaky_relu` is actually better (although not perfect due to underflow) since the exponential of that would be always no greater than 1.  The opposite is not since the result will easily overflow.

Reply:
But is it still correct? It seems to me that they change the activation function to `-leaky_relu`.

Reply:
The value range of the pre-softmax value doesn't matter.

Reply:
Hmm, I think `-leaky_relu` is just a little better than `leaky_relu` because the negative slope (the author used 0.2) is less steep than the positive slope (is 1.0). It can still go beyond 10 right?

Reply:
Oops I was thinking of `relu`, my bad.  Still, I think it will be hard to have the value explode in the negative case.